src/lava/magma/core/learning/learning_rule.py

```diff
@@ -0,0 +1,567 @@
+    impulse values by which to update traces upon spikes;
+    time constants by which to decay traces over time;
+    the length of the learning epoch;
+    a dict with dependencies as keys and the set of all traces appearing
+    with them in the specified learning rules as values;
+    the set of traces used by all specified learning rules.
+
+    From the user's perspective, a LoihiLearningRule object is to be used
+    as follows:
+    (1) Instantiate an implementation of LoihiLearningRule object with
+    learning rules given in string format for all three synaptic variables
+    (dw, dd, dt), as well as trace configuration parameters (impulse, decay)
+    for all available traces (x1, x2, y1), and the learning epoch length.
+
+    (2) The LoihiLearningRule object encapsulating learning-related information
+    is then passed to the Dense Process as instantiation argument.
+
+    (3) It will internally be used by ProcessModels to derive the operations
+    to be executed in the learning phase (Py and Nc).
+
+    Parameters
+    ----------
+
+    dw: str
+        ProductSeries representation of synaptic weight learning rule.
+    dd: str
+        ProductSeries representation of synaptic delay learning rule.
+    dt: str
+        ProductSeries representation of synaptic tag learning rule.
+    x1_impulse: float
+        Impulse by which x1 increases upon each pre-synaptic spike.
+    x1_tau: int
+        Time constant by which x1 trace decays exponentially over time.
+    x2_impulse: float
+        Impulse by which x2 increases upon each pre-synaptic spike.
+    x2_tau: int
+        Time constant by which x2 trace decays exponentially over time.
+    y1_impulse: float
+        Impulse by which y1 increases upon each post-synaptic spike.
+    y1_tau: int
+        Time constant by which y1 trace decays exponentially over time.
+    y2_impulse: float
+        Impulse by which y2 increases upon each post-synaptic spike.
+    y2_tau: int
+        Time constant by which y2 trace decays exponentially over time.
+    y3_impulse: float
+        Impulse by which y3 increases upon each post-synaptic spike.
+    y3_tau: int
+        Time constant by which y3 trace decays exponentially over time.
+    t_epoch: int
+        Duration of learning epoch.
+    rng_seed: int
+        Seed for the random number generators. If None, seed will be
+        set to current unix timestamp. Only used in fixed point
+        implementations.
+
+    """
+    def __init__(
+        self,
+        dw: ty.Optional[str] = None,
+        dd: ty.Optional[str] = None,
+        dt: ty.Optional[str] = None,
+        x1_impulse: float = 0.0,
+        x1_tau: int = 0,
+        x2_impulse: float = 0.0,
+        x2_tau: int = 0,
+        y1_impulse: float = 0.0,
+        y1_tau: int = 0,
+        y2_impulse: float = 0.0,
+        y2_tau: int = 0,
+        y3_impulse: float = 0.0,
+        y3_tau: int = 0,
+        t_epoch: int = 1,
+        rng_seed: int = None,
```

Comment:
These all seem to be `ty.Optional[float]` or `ty.Optional[int]`.

Reply:
done